### ISSUE TITLE
Specialize SpriteBatch Draw Methods

### ISSUE THREAD

**Issue by tomspilman** [MonoGame/MonoGame#5347]
While @nkast has done some great work with his low level optimizations there is still more to be done.

For example all of these functions in `SpriteBatch`:
```
void Draw(Texture2D texture, Vector2 position, Color color);
void Draw(Texture2D texture, Rectangle destinationRectangle, Color color);
void Draw (Texture2D texture, Vector2 position, Rectangle? sourceRectangle, Color color);
void Draw (Texture2D texture, Rectangle destinationRectangle, Rectangle? sourceRectangle, Color color);
```

All redirect to the same `DrawInternal` method which does a common set of optimizations.  We do this mainly to make the code easier to read and maintain as well as to void bugs.  This however comes at the expensive of performance.

Take `void Draw(Texture2D texture, Vector2 position, Color color)` for example.  This draw doesn't need rotation, doesn't need flipping of sprites, doesn't need UV calculations.  If this function created its own `SpriteBatchItem` internally it could do so much more efficiently than `DrawInternal` does now.

Same goes for the other specializations of `Draw` and `DrawFont` in `SpriteBatch`.  This matters when games are drawing 1000s for sprites per frame.

While this does mean a lot more duplicate code... it also means much more performance.  We can deal with the issue of duplicate code via more unit tests.  Performance here is more important.

**Comment by nkast** [comment]
I agree with all of the above. Actually I've done a few test recently to do exactly that. I pushed two more commits into #5267 the other day, the one called `Refactor SpriteBatchItem.Set()` is towards this goal. The performance improvements were negligible, the actual goal was to move calculations out of `.Set()`. 

> While this does mean a lot more duplicate code... [...]. We can deal with the issue of duplicate code via more unit tests. 

One pattern I notices was that we can redirect all `Draw(Texture2D texture, Vector2 position, ...)` into the similar `Draw(Texture2D texture, Rectangle destinationRectangle, ...)` without too much overhead. 

Another thing that became obvious during this fun process :dizzy_face: was the need for a standardized benchmark that can run a number of tests and compare them against the baseline.

**Comment by tomspilman** [comment]
> One pattern I notices was that we can redirect all `Draw(Texture2D texture, Vector2 position, ...)` into the similar `Draw(Texture2D texture, Rectangle destinationRectangle, ...)` without too much overhead.

Well our goal would be zero extra overhead.  I think sprite rendering performance is important enough that we would duplicate a function just to remove even a single instruction worth of unnecessary code.

I think we start by duplicating the code in `DrawInternal` to every single `Draw` in `SpriteBatch`.  Then we optimize them for each particular version of `Draw` and its arguments.  Then if we see two `Draw`s that are 100% identical we can consider sharing that code.

When you render 1000s of sprites a frame to hit 60fps... just a little less code matters.

It is less apparent on a modern 4.0 GHz PC...  but it certainly hurts lower end mobile devices and consoles.

> a standardized benchmark that can run a number of tests and compare them against the baseline.

I assume you mean performance here.  Yeah we do need some sort of basic performance benchmarks.

**Comment by nkast** [comment]
> Well our goal would be zero extra overhead. I think sprite rendering performance is important enough that we would duplicate a function just to remove even a single instruction worth of unnecessary code.

Sure. Although during my tests I found myself keep copying & moving code and though it might be helpful to have this redirection, at least in the early stages of refactoring & refining.

> I think we start by duplicating the code in DrawInternal to every single Draw in SpriteBatch. Then we optimize them for each particular version of Draw and its arguments. Then if we see two Draws that are 100% identical we can consider sharing that code.

For the sort version I implemented them from scratch, because I wanted to understand what exactly was going on. My thinking was that as we move to the more complex methods, the first code should be identical and just apply extra transformations. As I moved to the more complex methods I found that even the *position* methods need to keep a Rectangle to apply any transformations.
Bad news: I lost that branch because I never published it in github.

So, basically, I propose to implement the *destinationRectangle* methods first, starting from the short/simple ones. Once we are happy with the results we expand and optimize the *position* methods.

**Comment by nkast** [comment]
I can send a PR within the next week for the simple Draw Methods for further discussion.
 
@tomspilman Also I'd like your opinion on something like this:
https://github.com/MonoGame/MonoGame/compare/develop...nkast:tncSpriteBatchDXOpt
I am using a VertexBuffer and copy the Vertices directly into the mapped memory.
I'm trying to combine the two in this PR but still needs work and I'm not so sure about the results. The GL version will certainly take a hit in performance.
https://github.com/MonoGame/MonoGame/compare/develop...nkast:tncSpriteBatchMappedMemory

**Comment by vcrobe** [comment]
> When you render 1000s of sprites a frame to hit 60fps... just a little less code matters.

And of course this give you a little extra time in each frame to do a better AI.

**Comment by Jjagg** [comment]
> and the GL version will certainly take a hit in performance

@nkast  Do you mean if you make it use a VertexBuffer too? Why not make the changes to the DX implementation and keep things as is for GL? It's not like there's a lot of maintenance for SpriteBatch so having more code fragmentation for better performance seems like a nice tradeoff for something used so often. #speedoverbeauty

Great job on the optimizations you've done so far!

**Comment by tomspilman** [comment]
> I am using a VertexBuffer and copy the Vertices directly into the mapped memory. +12% perf.

I have a branch with just that same optimization from 2 years ago...  there is a complication with that.

Our OpenGL implementation of `DrawUserPrimitive` was actually faster than using a `DynamicVertexBuffer` when I tried this before.  In particular on iOS.

So we would need to maintain two paths to do that which sounds unappealing at the time.

Now we could add some sort of `SPRITEBATCH_USE_BUFFERS` macro around this new path with the fallback to `DrawUserPrimitive`.  Then on a platform-by-platform basis we define `SPRITEBATCH_USE_BUFFERS` if that is faster on this platform.

**Comment by KonajuGames** [comment]
We can't improve one at the expense of another, especially when the other
is the API used on mobile devices that are in most need of the performance
improvements.

**Comment by nkast** [comment]
Ok, I remembered now that I wanted to convert .Set() to accept a `ref rectangle destRect` instead of `float x, float y, float w,float h`. So it made sense at the time to reroute all `Draw(..., position, ...)` methods through `Draw(..., dest, ...)`, but the problem is that Position is float and Rectangle will convert it all to ints. 
so scrap that.

**Comment by tomspilman** [comment]
> Ok, I remembered now that I wanted to convert .Set() 

Really we should inline all of that.  I don't think there is any benefit to share any code in those `Draw` methods.  Inline all of it, remove any extra work that isn't required by that draw.

**Comment by nkast** [comment]
I have a branch ready where I refactored the simple versions of Draw().
I would like to wait for #5267 and move some of it's optimizations on this new branch for a fair comparison.
https://github.com/MonoGame/MonoGame/compare/develop...nkast:tncSpriteBatchDraw

> Really we should inline all of that. I don't think there is any benefit to share any code in those Draw methods. Inline all of it, remove any extra work that isn't required by that draw.

Some interesting findings : The inline version of `Draw(Texture2D,Vector2,Color)` is ~10% faster than rerouting it through a call to the new inlined `Draw(texture,position, null, color)`. That's way too high to ignore.

**Comment by amerkoleci** [comment]
I would anyway suggest thinking also on SpriteBatch instanced rendering, I saw an example of instanced sprite batch rendering under DirectX11.

**Comment by Jjagg** [comment]
> instanced rendering

I think that won't provide much benefit here because we're only drawing rectangles so the number of vertices is very low. Though we would transfer some minor computation to the GPU (rotation, scaling). This would be a nice experiment. Note that when a custom effect with a vertex shader is passed we can't do instanced rendering, so we would need to maintain two paths.

**Comment by LithiumToast** [comment]
Since we are on the topic of amending `SpriteBatch`'s API. What do you guys think of the following?
Support for passing in a model-to-world matrix by reference using `ref` when enqueuing a sprite. This matrix would be used to transform the 4 vertices of the quadrilateral representing a sprite from model space to world space. The reason I suggest this is because it is likely that the world transform of a sprite (position, rotation, origin, etc) doesn't change every frame and thus doesn't need to be recalculated every frame.

**Comment by willmotil** [comment]
I once argued to shawn hargraves this should be a internal thing for xna 5.0. Since SpriteBatch is not only aimed at new people but ease of use. Errr that i shouldn't need to create my own call to spin a image in place properly like so.

```
public static void DrawToScreenRects(Texture2D textureobj, Rectangle scrdraw_rect, Rectangle texturerect, float rot, float scale)
        {
            scrdraw_rect.Width = (int)(scrdraw_rect.Width * scale);
            scrdraw_rect.Height = (int)(scrdraw_rect.Height * scale);
            Vector2 toff = new Vector2(texture_offset.X + texturerect.Width * .5f, texture_offset.Y + texturerect.Height * .5f);
            BgEngine.Spritebatch.Draw(textureobj, scrdraw_rect, texturerect, SpriteColor, rot, toff, flip_effect, 0);
        }
```

**Comment by UncleThomy** [comment]
I've recently noticed that

https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/Effect/Resources/SpriteEffect.fx

makes a matrix transformation for each vertex. 

I would guess that in most cases this is not needed and should only occur when the user specifies a transformation. 

It *might* be worth considering having a second, simpler vertex shader which just passes through the input.

**Comment by LithiumToast** [comment]
@UncleThomy You mean [line 34](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/Effect/Resources/SpriteEffect.fx#L34)?

That transformation is transforming the vertex position with a WorldViewProjection matrix. See [here](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/Effect/SpriteEffect.cs#L73) and [here](https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/SpriteBatch.cs#L148). Kinda no way around that.

**Comment by nkast** [comment]
@UncleThomy 
That is both the user transformation and the screen projection.
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/SpriteBatch.cs#L146-L148
We can break up Begin() method to one that avoid the multiplication when user does not provide a Matrix. 

Can we really optimize the effect any more than than that? I was thinking that we could move more of the expensive operations (rotation & scale) into the effect, but is it possible without hurting the simple cases?

**Comment by tomspilman** [comment]
I've never seen the `SpriteBatch` vertex shader be a bottleneck in any game...  it does very little work.  That minuscule performance improvement is not worth the hassle of maintaining two versions of it.

The problems are all CPU side and mostly with our implementation of the API.

**Comment by theZMan** [comment]
There's too many variables at play at this point... remember MonoGame is cross platform so moving stuff from CPU to GPU might not work well on all platforms.. and waht if the spritebatch is being used in a 3d game which is already very GPU heavy... making general purpose frameworks is fun ;-)

So other than optimisations that clearly remove some processing its very hard to do this without cross platform profiling. Heck even CPU optimisations at the cost of memory can affect platforms with lower RAM so you can't even call those easy.

What we really need is some decently complex games that run on several platforms that we could profile and see exactly where we need to do work.

**Comment by UncleThomy** [comment]
i understand it's not worth it, the amount of vertex shader transformations is ridiculously low in comparison. Just a thought, shouldn't have written it.

@the other guys, the WVP trafo is pointless though, on CPU we transform backwards from aspect coordinates to normalized, just to trafo back on GPU.

**Comment by willmotil** [comment]
Well i been working on bypassing drawstring altogether on and off so i could wrap text so i have a prototype which makes it way easier to see everything that it does.

Let me summarize the case of drawstring for a moment.

Drawstring calls DrawInto after it accesses character source were it copied in the text... 
_(which btw is a redundant duncil structure.)_
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/SpriteBatch.cs#L146-L148

DrawInto in Spritefont does most of the work to decide were to place a glyph
https://github.com/MonoGame/MonoGame/blob/develop/MonoGame.Framework/Graphics/SpriteFont.cs

Drawinto then calls back to DrawInternal back in SpriteBatch. were it started out

So its flopping around all over but it can in total be reduced to the below.
A fonts glyph information, one method that takes a stringbuilder, and a call to draw internal.

This is what drawstring does simplified right up to were it hits draw internal but in a messy convoluted way. This code below will do exactly what DrawString does by just copy pasting it next to a game1.

```
    public class SpriteFontDraw
    {
        private Dictionary<char, SpriteFont.Glyph> _glyphs;
        private SpriteFont.Glyph defaultGlyph;
        private SpriteFont tsf;
        private char defaultfontchar = ' ';
        private static SpriteBatch spritebatch;
        private static Texture2D dottexture;
        // this is mine its new for buffering words to wrapp text directly within drawstring
        private DrawableWord word = new DrawableWord();

        public void LoadInitialize(SpriteBatch spriteBatch, SpriteFont spritefont)
        {
            SetSpriteFont(spritefont);
            spritebatch = spriteBatch;
            CreateDotTexture();
        }

        public void SetSpriteBatch(SpriteBatch spriteBatch)
        {
            spritebatch = spriteBatch;
            CreateDotTexture();
        }

        public void SetSpriteFont(SpriteFont s)
        {
            tsf = s;
            _glyphs = tsf.GetGlyphs();
            defaultGlyph = new SpriteFont.Glyph();
            if (tsf.DefaultCharacter.HasValue)
            {
                defaultfontchar = (char)(tsf.DefaultCharacter.Value);
                defaultGlyph = _glyphs[defaultfontchar];
            }
        }

        //
        // .... overloads and wrapps removed for brevity
        //

        public void DrawText(SpriteBatch spriteBatch, SpriteFont tsf, StringBuilder text, Vector2 position, Color color,
        float rotation, Vector2 origin, Vector2 scale, SpriteEffects effect, float depth, Rectangle textboundryarea)
        {
            var LineSpacing = (float)tsf.LineSpacing;
            var Spacing = tsf.Spacing;
            bool flip = false;
            bool flippedVert = false;
            bool flippedHorz = false;
            Vector2 msize = Vector2.Zero;
            Vector2 offset = Vector2.Zero;
            Vector2 q = new Vector2((float)(Math.Sin(rotation)), (float)(Math.Cos(rotation)));
            Rectangle dest = new Rectangle();
            var currentGlyph = defaultGlyph;
            var firstGlyphOfLine = true;

            // for now ill set a flag if we cannot skip all the spriteeffect calculations
            // flipping is a pretty big punishment to this method
            if (effect != SpriteEffects.None)
            {
                flip = true;
                // measure string needs fixed next or to be gotten rid off or around, its terrible.
                // maybe a entirely different approach to measuring 
                // were we get a pre calculated measured drawable buffer back instead
                msize = tsf.MeasureString(text);
                flippedHorz = (effect & SpriteEffects.FlipHorizontally) == SpriteEffects.FlipHorizontally;
                flippedVert = (effect & SpriteEffects.FlipVertically) == SpriteEffects.FlipVertically;
            }
            for (var i = 0; i < text.Length; i++)
            {
                var c = text[i];
                if (c == '\r')
                    continue;
                if (c == '\n')
                {
                    offset.X = 0;
                    offset.Y += LineSpacing;
                    firstGlyphOfLine = true;
                    continue;
                }

                // needs to be touched up
                if (_glyphs.ContainsKey(c))
                    currentGlyph = _glyphs[c];
                else
                    if (!tsf.DefaultCharacter.HasValue)
                        throw new ArgumentException("Text Contains a Unresolvable Character");
                    else
                        currentGlyph = defaultGlyph;

                // Solves the problem- the first character on a line with a negative left side bearing.
                if (firstGlyphOfLine)
                {
                    offset.X = Math.Max(currentGlyph.LeftSideBearing, 0);
                    firstGlyphOfLine = false;
                }
                else
                    offset.X += Spacing + currentGlyph.LeftSideBearing;

                // matrix calculations unrolled varys max 8 mults and up to 10 add subs.
                var m = offset;
                m.X += currentGlyph.Cropping.X;
                m.Y += currentGlyph.Cropping.Y;

                // oh still more sprite effect penaltys good need a no sprite effect version for sure
                if (flip) 
                {
                    if (flippedHorz)
                        m.X = msize.X - (m.X + currentGlyph.BoundsInTexture.Width);
                    if (flippedVert)
                        m.Y = msize.Y - (m.Y + currentGlyph.BoundsInTexture.Height);
                }
                var r = m - origin;
                var p = r;
                if (rotation != 0)
                {
                    p.X = r.X * q.Y - r.Y * q.X;
                    p.Y = r.X * q.X + r.Y * q.Y;
                }
                dest = new Rectangle(
                    (int)(p.X * scale.X + position.X),
                    (int)(p.Y * scale.Y + position.Y),
                    (int)(currentGlyph.BoundsInTexture.Width * scale.X),
                    (int)(currentGlyph.BoundsInTexture.Height * scale.Y)
                    );
                    spriteBatch.Draw(tsf.Texture, dest, currentGlyph.BoundsInTexture, color, rotation, Vector2.Zero, effect, depth);
                offset.X += currentGlyph.Width + currentGlyph.RightSideBearing;
            }
        }
}
```

I have been working on and off on this for some time now im just bored quickly and easily distracted, Its not yet done though it works its not optimized at all really but its good for looking at whats going on.

Though i did this for simplifying text wrapping those methods are omitted for clarity.

**Comment by MichaelDePiazzi** [comment]
Just a thought... instead of duplicating a lot of code here, perhaps just using the `[MethodImpl(MethodImplOptions.AggressiveInlining)]` attribute on relevant methods and doing some refactoring would be enough to get the compiler to do all this inlining for you? (Further reading: http://stackoverflow.com/a/8746128)

I think it's at least worth investigating. If this does the trick, then it would mean you get both maintainable and high performance code. If I get some time, I'll try messing around with this.

**Comment by MichaelDePiazzi** [comment]
FYI - I've been messing a little with the aggressive inlining option and have had some success with it. Some conditions will cause inlining to fail though (e.g. too many parameters, modifying parameter values, etc) which can be tricky to diagnose.

[BenchmarkDotNet](http://benchmarkdotnet.org/) has been very useful in regards to this though. Not only is it great for micro-benchmarks, it also has a [JIT Inlining Events](http://benchmarkdotnet.org/Configs/Diagnosers.htm) diagnoser which can tell you the reason why inlining has failed. It also supports running benchmarks under different environments so you can test how code performs using different JIT compilers and runtimes.

I did an experiment using the simplest Draw method (i.e. texture, position, color overload) as this would be the best case scenario for inlining. With inlining, it took approximately 60% less time than the original version!

For comparison, I did a manually inlined version (i.e. all code within 1 method) and a compiler inlined version (i.e. several small helper methods using aggressive inlining). The performance was practically identical, and in fact the compiler inlined version seemed to perform just slightly better.

I now definitely think the "aggressive inlining" option is worth pursuing as it means the code will be a lot easier to read and maintain while still performing well.

Thoughts?

P.S. Here's the output from BenchmarkDotNet:

``` ini
BenchmarkDotNet=v0.10.1, OS=Microsoft Windows NT 6.2.9200.0
Processor=AMD Phenom(tm) II X4 955 Processor, ProcessorCount=4
Frequency=3138888 Hz, Resolution=318.5842 ns, Timer=TSC
  [Host]     : Clr 4.0.30319.42000, 32bit LegacyJIT-v4.6.1586.0
  Job-FMXRTV : Clr 4.0.30319.42000, 32bit LegacyJIT-v4.6.1586.0

EvaluateOverhead=True  MaxStdErrRelative=0.001  RemoveOutliers=True  
Jit=LegacyJit  Platform=X86  Runtime=Clr  
Force=True  
```
Method | Mean | StdDev | StdErr | Min | Q1 | Median | Q3 | Max | P95 | Op/s | Scaled | Scaled-StdDev
--- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | ---
Original | 69.6027 ns | 0.0420 ns | 0.0121 ns | 69.5304 ns | 69.5741 ns | 69.5985 ns | 69.6438 ns | 69.6587 ns | 69.6571 ns | 14367268.17 | 1.00 | 0.00
Optimised | 25.5017 ns | 0.0027 ns | 0.0008 ns | 25.4987 ns | 25.4994 ns | 25.5016 ns | 25.5032 ns | 25.5084 ns | 25.5060 ns | 39212995.66 | 0.37 | 0.00
ManualInline | 27.7248 ns | 0.1068 ns | 0.0296 ns | 27.5813 ns | 27.6443 ns | 27.7049 ns | 27.7772 ns | 27.9421 ns | 27.9224 ns | 36068824.84 | 0.40 | 0.00

And here's my work in progress branch with the changes I've made:
https://github.com/MichaelDePiazzi/MonoGame/compare/finaldays...MichaelDePiazzi:fdSpriteBatchOpt2?w=1
(NOTE: The "Optimised" method is named `DrawOpt2` and the "ManualInline" method is named `DrawOpt3`)

**Comment by theZMan** [comment]
I assume you are only testing on windows though we have no idea how well inlining works on other platforms/compilers.

Manual inlining works on all platforms, using the compiler doesn't necessarily do the same.

**Comment by MichaelDePiazzi** [comment]
@theZMan - The above benchmarks were run using the LegacyJIT x86 compiler and the CLR runtime. BenchmarkDotNet does support other environments (e.g. RyuJIT, x64, Mono), so I know it at least covers desktop platforms. But I don't know about other platforms (e.g. mobile, console).

I do know manual inlining is a safer bet than asking the compiler to do it. But it will also result in the code being more complex and harder to maintain. So I was just trying to investigate a possible alternative solution before we embark on this. Using compiler inlining means we can write simpler and easier to maintain code. But the downside is that we'd have to benchmark on all platforms to ensure it's actually being inlined on each platform.

Anyways, just some food for thought fwiw. There are definitely pros and cons to both approaches. But I'd definitely understand if the consensus was to use manual inlining.

**Comment by tomspilman** [comment]
> I do know manual inlining is a safer bet than asking the compiler to do it. But it will also result in the code being more complex and harder to maintain. 

I 100% agree that manual inlining sucks and is harder than normal to maintain.  I am the first one to suggest breaking up code for reusability and easy maintenance.

However the behavior of C# compilers/jits/aots is all over the map still.  We don't know what we will get from inline attributes.  Some platforms it might work and we get the expected performance and others might not inline at all...  Something difficult to spot and maintain as we add platforms.

So to me it is less maintenance to manually inline in these performance critical areas.  We can also lessen the costs by using unit tests to ensure we don't break behavior.

**Comment by MichaelDePiazzi** [comment]
@tomspilman - That's totally reasonable. It was an interesting exercise anyway playing around with this inlining option as well as BenchmarkDotNet. I've already used it to optimise some other performance critical code of mine, so it has been beneficial for me at least. I only have to worry about Windows though... ;)

**Comment by Vusroda** [comment]
Are you going with this? My game would benefit from this optimization :)

**Comment by MMKnight** [comment]
Manual optional argument is inconvienent. I have to create new method instead of writing lots of duplication.

**Comment by Jjagg** [comment]
@MMKnight I don't understand what you're saying. Manual optional argument?

> I have to create new method instead of writing lots of duplication.

Also don't know what you mean by this, byt sounds like a good thing?

Anyway, as a user of MonoGame, this issue doesn't affect you (noticeably), because it's not related to the API.

**Comment by MMKnight** [comment]
@Jjagg I have to use ton of optional arguments to make it obsoleted.
